MultiRobotControl/ay22_gdk
# Mh/feature/matlab lidar tracking

Matlab Lidar Tracking

To test:
git fetch
git checkout mh/feature/matlab_lidar_tracking

To run:
- Modify /rabbit.urdf file and change marker radius to 5.0
- From the terminal window:
roslaunch mrc_examples marina_1x.launch 
- In a new terminal window: 
roslaunch mrc_examples rviz_cora1.launch
- In a third terminal window:
open matlab from the /matlab directory

First run lidar_dist_head to verify lidar is functioning

- Most likely, the following message will be displayed in the matlab command window:
Error using rosReadXYZ
Expected input to be one of these types:
struct
Instead its type was double.
Error in rosReadXYZ (line 46)
    validateattributes(msg,{'struct'},{'scalar'},'rosReadXYZ');
Error in pcloud2dist (line 8)
xyz = rosReadXYZ(lidar);
Error in lidar_dist_head (line 30)
    [dist, psi] = pcloud2dist(PCLOUD);

- I ran into a similar issue with hw7, where the vbap_single file threw an error the first time I tried to run it.
- If you run lidar_dist_head again, it should begin to display distances and heading angles in the command window:
Range_GPSRange =
  1×2 single row vector
  109.8332  113.2021
Psi_GPSPsi =
  1×2 single row vector
  115.7416  116.0469

- The first value in each row is the estimated range/psi based on lidar measurements. As a comparison, the second value is the range/psi using GPS. If the rabbit is out of range for lidar, the range and psi values will read NaN
- ctrl C to terminate
- Now run lidar_tracking. For troubleshooting, this file will display the linear and angular command velocities in the command window as the file runs:
v_c =
     0
r_c =
     0
v_c =
  single
   12.9807
r_c =
  single
    0.9748

- To avoid issues when the target is out of range, the pcloud2cmd_vel function will set both command velocities to zero when it recieves NaN for range/psi. This will stop the boat while it waits for the target to come back into range. This should only happen if the lidar_tracking script is started when the target is out of range.
